### PR TITLE
[Collaborative_optimizations] Cluster-preserving quantization aware training

### PR DESCRIPTION
**Pull request by Ruomei**
This PR adds Cluster-preserving Quantization Aware Training (CQAT), which injects clustering trainable variables and corresponding TensorFlow ops into the original QAT Keras graph so that they could be trained together. In this way, the nice distribution of clustered weights can be kept after QAT, which gives benefits for model accuracy and compression rate.

Initial results:

| Model | Items | Baseline | Clustered Model | QATed Model | CQATed Model | 
| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
| Mobilenet_v1 on CIFAR-10 | FP32 Top1 Accuracy | **94.88%**  | 94.48%  | (Fake INT8) 94.80% | (Fake INT8) 94.60% |
|                        | INT8 full integer quantization | 94.65% | **94.41%** | 94.77% | **94.52%** |
|                        |                                                  | 3.00 MB | 2.00 MB | 2.84 MB | 1.94 MB |
| Mobilenet_v1 on ImageNet | FP32 Top 1 Accuracy | 71.07% | 65.30% |  (Fake INT8) 70.39% | (Fake INT8) 65.35% |
|                        | INT8 full integer quantization | 69.34% | **60.60%** | 70.35% | **65.42%** |
|                        | Compression                           | 4665568 -> 3886277 | 4665568 -> 3035752 | 4569416 -> 3804871 | 4569472 -> 2912655|

### CODE REVIEW

**Comment by Ruomei** [comment]
Hi, @daverim @nutsiepully  @akarmi @wwwind, could you please review this PR?

File: tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/cluster_preserve_quantize_registry_test.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+  class CustomLayer(layers.Layer):
+    # a simple custom layer with training weights
+    def __init__(self):
+      super(ClusterPreserveQuantizeRegistryTest.CustomLayer, self).__init__()
+
+    def build(self, input_shape=(2, 2)):
+      self.add_weight(shape=input_shape,
+                      initializer="random_normal",
+                      trainable=True)
+
```
Comment (daverim) [inline tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/cluster_preserve_quantize_registry_test.py:65]: Should be a double-space after classes

File: tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/cluster_preserve_quantize_registry_test.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+      self.cluster_preserve_quantize_registry.\
+        apply_cluster_preserve_quantize_config(
+          self.layer_conv2d, self.CustomQuantizeConfig)
+
+    with self.assertRaises(ValueError,
+                           msg="Unregistered layers should raise error."):
+      self.cluster_preserve_quantize_registry.\
+        apply_cluster_preserve_quantize_config(
+          self.layer_custom, self.CustomQuantizeConfig)
+
```
Comment (daverim) [inline tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/cluster_preserve_quantize_registry_test.py:118]: Ditto

File: tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/cluster_preserve_quantize_registry.py [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+      # Embedding need verify from 8bit qat
+      # layers.Embedding: _ClusterPreserveInfo(['embeddings'], []),
+  }
+
+  _DISABLE_CLUSTER_PRESERVE = {
+      layers.DepthwiseConv2D,
+  }
+
+  def __init__(self):
+
```
Comment (daverim) [inline tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/cluster_preserve_quantize_registry.py:121]: Delete extra space

File: tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/cluster_preserve_quantize_registry.py [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+    """Returns whether the registry supports this layer type.
+
+    Args:
+      layer: The layer to check for support.
+
+    Returns:
+      True/False whether the layer type is supported.
+
+    """
+
```
Comment (daverim) [inline tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/cluster_preserve_quantize_registry.py:166]: For all these

File: tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/cluster_preserve_quantize_registry.py [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+  def _disable_cluster_preserve(cls, layer):
+    """Returns whether disable this layer for preserving clusters.
+
+    Args:
+      layer: The layer to check for disabling.
+
+    Returns:
+      True/False whether disabling this layer for preserving clusters.
+    """
+
```
Comment (daverim) [inline tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/cluster_preserve_quantize_registry.py:152]: Ditto

**Review by daverim — APPROVED** [review]
LGTM, some small nits though about formatting / whitespace

File: tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/cluster_preserve_quantize_registry_test.py (hunk 2 above)
Comment (Ruomei) [reply tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/cluster_preserve_quantize_registry_test.py:118]: Thanks, I found similar problems in other places of the files and made changes too.

File: tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/cluster_preserve_quantize_registry_test.py (hunk 1 above)
Comment (Ruomei) [reply tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/cluster_preserve_quantize_registry_test.py:65]: For this one, I am not sure. Does the inner class (not a top-level definition) follow the same rule of 2 blank lines? I checked the coding style given to us before, and it says:

> Two blank lines between top-level definitions, be they function or class definitions. One blank line between method definitions and between the class line and the first method. No blank line following a def line. Use single blank lines as you judge appropriate within functions or methods.

Is there any other coding style you are using internally atm? Also, I used the pylint [here ](https://www.tensorflow.org/community/contribute/code_style) and it never reported issues of lines or spaces. Therefore, I switched to `pycodestyle`. Is there a tool integrated with your current coding style for TFMOT that we could use in the future?

File: tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/cluster_preserve_quantize_registry.py (hunk 3 above)
Comment (Ruomei) [reply tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/cluster_preserve_quantize_registry.py]: Done, thanks.

File: tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/cluster_preserve_quantize_registry.py (hunk 4 above)
Comment (Ruomei) [reply tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/cluster_preserve_quantize_registry.py]: Done, thanks.

File: tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/cluster_preserve_quantize_registry.py (hunk 5 above)
Comment (Ruomei) [reply tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/cluster_preserve_quantize_registry.py]: Done, thanks.

**Review by akarmi — APPROVED** [review]

**Review by daverim — APPROVED** [review]
